### PR TITLE
indexheader.Reader implementation that runs in a thread pool

### PR DESCRIPTION
**Pull request by 56quarters**
This adds a new Thanos indexheader.Reader implementation that uses
a dedicated pool of OS threads (via `runtime.LockOSThread()`) to
ensure that mmap page faults (caused by reading index files from
disk using mmap) do not block a thread used for scheduling other
goroutines.

Signed-off-by: Nick Pillitteri <nick.pillitteri@grafana.com>

#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/storegateway/indexheader/threadpool.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	pool       chan *OSThread
+	numThreads int
+}
+
+func NewThreadPool(ctx context.Context, cancel context.CancelFunc, num int) (*Threadpool, error) {
+	if num <= 0 {
+		return nil, nil
+	}
+
+	if num >= runtime.GOMAXPROCS(0) {
```
Comment (johannaratliff) [inline pkg/storegateway/indexheader/threadpool.go:26]: To clarify - this is to check if the threadpool is going to be >= number of CPUs that can be executing simultaneously, yes?

**Review by johannaratliff — COMMENTED** [review]
To the best of my understanding, this lgtm

File: pkg/storegateway/indexheader/threadpool.go (hunk 1 above)
Comment (56quarters) [reply pkg/storegateway/indexheader/threadpool.go:26]: Correct. If the size of the threadpool is too large the service won't be able to run anything since the pool makes exclusive use of N threads.

**Comment by Logiraptor** [comment]
If I understand correctly, the root problem is that Go's scheduler doesn't know that some memory access is actually blocking (due to mmap page faults). Therefore it can't smartly schedule more work from other waiting goroutines in the meantime, like it would for normal blocking I/O. Since Go limits the number of running goroutines to GOMAXPROCS, there is a problem when GOMAXPROCS mmap page faults occur at the same time. Effectively the whole system blocks waiting for disk when it could proceed with other work. 

This change forces all usage of the indexheader.Reader to run in a set of N OS threads. If you set N == GOMAXPROCS, you're effectively back to the same problem as before, correct? The scheduler won't make any distinction between thread pool goroutines and others, so you can still have the system lock up as before - but now there's more overhead in all the synchronization going on. The benefit is only realized when N < GOMAXPROCS, because you are "reserving" some portion of (GOMAXPROCS - N) threads that you know can _never_ end up blocked by mmap. The (N+1)th goroutine that attempts to use  indexheader.Reader will be blocked on the threadpool's channel instead of mmap, which allows the scheduler to make a smart decision and schedule other work.

So then the downsides of this approach would be two things: 1) all accesses if the indexheader.Reader now require synchronization to claim one of N threads and 2) even in the best case scenario where no page faults are necessary, we still throttle reads.

Is that right?

Pretty neat idea!

**Comment by 56quarters** [comment]
> If I understand correctly, the root problem is that Go's scheduler doesn't know that some memory access is actually blocking (due to mmap page faults). Therefore it can't smartly schedule more work from other waiting goroutines in the meantime, like it would for normal blocking I/O. Since Go limits the number of running goroutines to GOMAXPROCS, there is a problem when GOMAXPROCS mmap page faults occur at the same time. Effectively the whole system blocks waiting for disk when it could proceed with other work.

Correct!

> This change forces all usage of the indexheader.Reader to run in a set of N OS threads. If you set N == GOMAXPROCS, you're effectively back to the same problem as before, correct? The scheduler won't make any distinction between thread pool goroutines and others, so you can still have the system lock up as before - but now there's more overhead in all the synchronization going on. The benefit is only realized when N < GOMAXPROCS, because you are "reserving" some portion of (GOMAXPROCS - N) threads that you know can _never_ end up blocked by mmap. The (N+1)th goroutine that attempts to use indexheader.Reader will be blocked on the threadpool's channel instead of mmap, which allows the scheduler to make a smart decision and schedule other work.

If you set N == GOMAXPROCS I suspect the service wouldn't even be able to start up but yes, otherwise you are correct. We force callers to block on the threadpool channel instead of mmap so that the service doesn't become unresponsive.   

> So then the downsides of this approach would be two things: 1) all accesses if the indexheader.Reader now require synchronization to claim one of N threads and 2) even in the best case scenario where no page faults are necessary, we still throttle reads.
> 

Correct, all index file accesses now going through this threadpool even when memory wouldn't need to be paged in. If we wanted to be fancy we could automatically grow or shrink the pool but I didn't want to complicate this without knowing if it improves things. We'll have to tune the number of threads in the pool based on running this but I don't _think_ there's much harm in sizing it such that we almost never block waiting for a thread. I suppose I should instrument this in some way so we can tell how many threads are typically in use.

**Comment by Logiraptor** [comment]
> I suppose I should instrument this in some way so we can tell how many threads are typically in use.

Maybe something similar to the way [query time is measured in stages](https://github.com/grafana/mimir/blob/25634ff2d68697f7b52ad0a94d4bbd7b2b9fe479/vendor/github.com/prometheus/prometheus/promql/engine.go#L305-L308)?

For a given call to the threadpool, how long is spent blocked on synchronization vs actually running.

File: pkg/storegateway/bucket_stores.go [hunk 2]
```diff
@@ -144,13 +151,27 @@ func NewBucketStores(cfg tsdb.BlocksStorageConfig, shardingStrategy ShardingStra
 		return nil, errors.Wrap(err, "create chunks bytes pool")
 	}
 
+	// TODO(56quarters): Make this configurable
+	if u.threadPool, err = mimir_indexheader.NewThreadPool(context.Background(), nil, 1); err != nil {
```
Comment (colega) [inline pkg/storegateway/bucket_stores.go:155]: Why don't we pass `runtime.GOMAXPROCS()` here instead of `1`?

(Also related question below about why it should be `runtime.GOMAXPROCS()-1`, but let's keep that in a separate thread)

File: pkg/storegateway/indexheader/threadpool.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+		for i := 0; i < t.numThreads; i++ {
+			thread := <-t.pool
+			thread.Join()
+		}
+	}
+}
+
+func (t *Threadpool) Start() {
+	go t.start()
+}
```
Comment (colega) [inline pkg/storegateway/indexheader/threadpool.go:61]: `Start` isn't called anywhere.

File: pkg/storegateway/indexheader/thread.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+			o.localTasks.Dec()
+			o.res <- execResult{value: val, err: err}
+			break
+		}
+	}
+}
+
+func (o *OSThread) Start() {
+	go o.start()
+}
```
Comment (colega) [inline pkg/storegateway/indexheader/thread.go:60]: To my best understanding, this is ensures that we run locked to a single OS thread, but there's nothing ensuring (or even trying) that `OSThread`s in the `ThreadPool` are running on _different_ threads. All OSThread workers could be scheduled on the same thread even if GOMAXPROCS > 1.

Comment (colega) [reply pkg/storegateway/indexheader/thread.go:60]: Deleting this thread because I've read the docs again:
```
// The calling goroutine will always execute in that thread,
// and no other goroutine will execute in it,
```

File: pkg/storegateway/indexheader/threadpool.go (hunk 1 above)
Comment (colega) [reply pkg/storegateway/indexheader/threadpool.go:26]: I think we should remove this limitation:

> If the size of the threadpool is too large the service won't be able to run anything since the pool makes exclusive use of N threads.

GOMAXPROCS != maximum number of threads our app can run. GOMAXPROCS is the number of threads where golang schedulers run (TIL this).

TL;DR, you can run 2*GOMAXPROCS thread-locked goroutines, and then run another one: https://go.dev/play/p/WFV3tysb4S5

File: pkg/storegateway/indexheader/threadpool.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+func (t *Threadpool) start() {
+	for range t.ctx.Done() {
+		for i := 0; i < t.numThreads; i++ {
+			thread := <-t.pool
+			thread.Join()
+		}
+	}
+}
```
Comment (colega) [inline pkg/storegateway/indexheader/threadpool.go:57]: Can you explain what this does?

As far as I understand, if you call start on a fresh Threadpool:
It pulls each thread from the pool, and calls `Join` on it, which is:
```go
func (o *OSThread) Join() {
	for range o.ctx.Done() {
		tasks := o.localTasks.Load()
		if tasks == 0 {
			break
		}
	}

	o.cancel()
}
```

Since there are no tasks, it will break on the first iteration of the loop, and cancel the local context and return.

Once this is done, the thread is stopped (because context is canceled) and nobody will put more threads in the pool (the only way of putting threads is in New() or by returning a thread from a call, but we have _joined_ that one on the first iteration)

File: pkg/storegateway/indexheader/threadpool.go (hunk 1 above)
Comment (56quarters) [reply pkg/storegateway/indexheader/threadpool.go:26]: Nice! TIL

File: pkg/storegateway/bucket_stores.go (hunk 2 above)
Comment (56quarters) [reply pkg/storegateway/bucket_stores.go:155]: `1` is just for while this is a POC, I didn't want to go to the trouble of wiring in the config required. I'm not sure what a reasonable value would be for this so I think we'll have to experiment. Though, to your point about GOMAXPROCS maybe it doesn't matter since the threads in this pool will either completely immediately or be blocked on a page fault.

File: pkg/storegateway/indexheader/threadpool.go (hunk 3 above)
Comment (56quarters) [reply pkg/storegateway/indexheader/threadpool.go:61]: Whoops, must of have lost that during a refactor

File: pkg/storegateway/indexheader/threadpool.go (hunk 5 above)
Comment (56quarters) [reply pkg/storegateway/indexheader/threadpool.go:57]: `ctx.Done()` is a channel so the loop body is blocked and doesn't run until the context is cancelled.

Comment (56quarters) [reply pkg/storegateway/indexheader/threadpool.go:57]: The use of child context here is confusing though. I'm going to work on making this a bit more straightforward.

File: pkg/storegateway/indexheader/threadpool.go (hunk 1 above)
Comment (colega) [reply pkg/storegateway/indexheader/threadpool.go:26]: (Me too :joy:)

File: pkg/storegateway/indexheader/threadpool.go (hunk 5 above)
Comment (colega) [reply pkg/storegateway/indexheader/threadpool.go]: Thank you. And yes, I misinterpreted the loop on ctx.Done() :facepalm:

**Comment by colega** [comment]
Overall this looks good to me. Pending to remove the `n <= GOMAXPROCS` check, and please make the number of threads in the pool configurable before merging.

---

OTOH, a nitpick, I'd consider [making the nil value of ThreadPool useful](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=385s), like:

```go
func (t *Threadpool) Start() {
	if t == nil {
		return
	}
	// ...
}

func (t *Threadpool) StopAndWait() {
	if t == nil {
		return
	}
	// ...
}

func (t *Threadpool) Call(fn func() (interface{}, error)) (interface{}, error) {
	if t == nil {
		return fn()
	}
	// ...
}
```

This way the caller (bucket.go) doesn't need to check whether the `ThreadPool` was nil or not, it would use whatever `NewThreadPool` has returned, even if it was the nil value because `n == 0`.

File: pkg/storegateway/bucket_stores.go (hunk 2 above)
Comment (colega) [reply pkg/storegateway/bucket_stores.go]: (After reading on how this works, passing in GOMAXPROCS also doesn't make sense).

Anyway 1 is okay to check that this works, but I think it will drastically decrease the performance.

**Comment by colega** [comment]
BTW I think it would be complicated to find a thread number for each use case (it even depends on the disk speed), I would consider adding an auto-scale for the threads: creating one if there's a caller waiting for... 100ms? And recycling them if (the ones on top of a minimum) haven't had a job during the last minute.

**Comment by 56quarters** [comment]
> BTW I think it would be complicated to find a thread number for each use case (it even depends on the disk speed), I would consider adding an auto-scale for the threads: creating one if there's a caller waiting for... 100ms? And recycling them if (the ones on top of a minimum) haven't had a job during the last minute.

Not a bad idea (I think this is roughly how "cached thread pools" in Java work) but I don't think I'm going to implement it in an effort to prevent this code from getting even more complicated than it already is. My guess (not verified yet) is that having any number of dedicated threads will be sufficient to mitigate the worst of the performance degradation we're seeing by preventing unrelated goroutines from stalling.

File: pkg/storegateway/indexheader/thread.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+		tasks := o.localTasks.Load()
+		if tasks == 0 {
+			break
+		}
+	}
+}
+
+func (o *OSThread) Call(fn func() (interface{}, error)) (interface{}, error) {
+	select {
+	case <-o.stopping:
```
Comment (colega) [inline pkg/storegateway/indexheader/thread.go:93]: One last nit, not blocking, I think we can make this:

```suggestion
func (o *OSThread) Call(ctx context.Context, fn func() (interface{}, error)) (interface{}, error) {
	select {
	case <-ctx.Done():
		return nil, ctx.Err()
	case <-o.stopping:
```

So the caller don't need to wait to grab a worker if the context is already canceled.

File: pkg/storegateway/indexheader/thread.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+import (
+	"runtime"
+
+	"go.uber.org/atomic"
+)
+
+type execResult struct {
+	value     interface{}
+	err       error
+	populated bool
```
Comment (pracucci) [inline pkg/storegateway/indexheader/thread.go:17]: Who use this? I can't find in which condition would be `false`

File: pkg/storegateway/indexheader/thread.go [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+func (o *OSThread) Stop() {
+	close(o.stopping)
+}
+
+func (o *OSThread) Join() {
+	for range o.stopped {
+		tasks := o.localTasks.Load()
+		if tasks == 0 {
+			break
+		}
```
Comment (pracucci) [inline pkg/storegateway/indexheader/thread.go:87]: Could you add a comment to explain why we need to check this? I'm not 💯 sure I understand the reason.

File: pkg/storegateway/indexheader/threaded_binary_reader.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+func (t *threadedReader) PostingsOffset(name string, value string) (index.Range, error) {
+	val, err := t.pool.Call(func() (interface{}, error) {
+		return t.reader.PostingsOffset(name, value)
+	})
+
+	if err != nil {
+		return index.Range{}, err
+	}
```
Comment (pracucci) [inline pkg/storegateway/indexheader/threaded_binary_reader.go:49]: [nit] I think this can be removed.

File: pkg/storegateway/indexheader/threaded_binary_reader.go [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+func (t *threadedReader) LookupSymbol(o uint32) (string, error) {
+	val, err := t.pool.Call(func() (interface{}, error) {
+		return t.reader.LookupSymbol(o)
+	})
+
+	if err != nil {
+		return "", err
+	}
```
Comment (pracucci) [inline pkg/storegateway/indexheader/threaded_binary_reader.go:61]: [nit] I think this can be removed.

File: pkg/storegateway/indexheader/threaded_binary_reader.go [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+func (t *threadedReader) LabelValues(name string) ([]string, error) {
+	val, err := t.pool.Call(func() (interface{}, error) {
+		return t.reader.LabelValues(name)
+	})
+
+	if err != nil {
+		return nil, err
+	}
```
Comment (pracucci) [inline pkg/storegateway/indexheader/threaded_binary_reader.go:73]: [nit] I think this can be removed.

File: pkg/storegateway/indexheader/threaded_binary_reader.go [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+func (t *threadedReader) LabelNames() ([]string, error) {
+	val, err := t.pool.Call(func() (interface{}, error) {
+		return t.reader.LabelNames()
+	})
+
+	if err != nil {
+		return nil, err
+	}
```
Comment (pracucci) [inline pkg/storegateway/indexheader/threaded_binary_reader.go:85]: [nit] I think this can be removed.

File: pkg/storegateway/indexheader/threadpool.go [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+	"errors"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+const (
+	LabelWaiting  = "waiting"
+	LabelComplete = "complete"
```
Comment (pracucci) [inline pkg/storegateway/indexheader/threadpool.go:18]: [nit] No need to expose them.

**Review by pracucci — COMMENTED** [review]
Neat idea! I like this idea and I wish it proves to be successful in your tests.

I think there's a missing piece of this work: when we create a thread pool, I think we should increase `GOMAXPROCS()` by the number of threads (and decrease it by the same number when it gets stopped). Reason is that threads in the thread pool are locked, and so won't be used by other goroutines. Enabling this feature we effectively reduce the actual number of threads available for the rest of the application. If we increase `GOMAXPROCS()` accordingly, we'll remove this side effect.

File: pkg/storegateway/indexheader/thread.go [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+func (o *OSThread) Start() {
+	go o.start()
+}
+
+func (o *OSThread) Stop() {
+	close(o.stopping)
+}
+
+func (o *OSThread) Join() {
+	for range o.stopped {
```
[the remark this answers is not in the document]
Comment (56quarters) [reply pkg/storegateway/indexheader/thread.go:83]: Yes, good point.

File: pkg/storegateway/indexheader/thread.go (hunk 7 above)
Comment (56quarters) [reply pkg/storegateway/indexheader/thread.go:17]: Exactly! It will never be `false` except for the zero value of this struct. When an `OSThread` is stopped the `res` channel is closed and so any further reads from it will result in the zero value for `execResult`. The code doing the read then checks for this to return the appropriate error when the channel is closed (and hence the thread pool is shutting down).

File: pkg/storegateway/indexheader/threadpool_test.go [hunk 15]
```diff
@@ -0,0 +1,3 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package indexheader
```
[the remark this answers is not in the document]
Comment (56quarters) [reply pkg/storegateway/indexheader/threadpool_test.go:3]: Yes! Will do

File: pkg/storegateway/indexheader/thread.go (hunk 6 above)
Comment (56quarters) [reply pkg/storegateway/indexheader/thread.go:93]: I don't think this is really possible. The `indexheader.Reader` interface doesn't take context parameters so there's no context to really provide to the method.

File: pkg/storegateway/indexheader/thread.go (hunk 8 above)
Comment (56quarters) [reply pkg/storegateway/indexheader/thread.go:87]: Actually, I don't think this needs to exist at all anymore. Since it's attempting a read from `o.stopped` it's guaranteed that any tasks run by the thread have already completed (because of the `defer func() { close(o.stopped) }()` in `.start()`

File: pkg/storegateway/indexheader/threadpool.go [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+	defer func() {
+		close(t.stopped)
+	}()
+
+	// The .stopping channel is never written so this blocks until the channel is
+	// closed at which point the threadpool is shutting down, so we want to stop
+	// each of the expected threads in it.
+	<-t.stopping
+	for i := 0; i < t.numThreads; i++ {
+		thread := <-t.pool
```
[the remark this answers is not in the document]
Comment (56quarters) [reply pkg/storegateway/indexheader/threadpool.go]: Fixed by keeping a slice in addition to the channel.

File: pkg/storegateway/indexheader/thread.go (hunk 6 above)
Comment (colega) [reply pkg/storegateway/indexheader/thread.go:89]: Oh I see. Okay.

File: pkg/storegateway/indexheader/thread.go [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+	}()
+
+	for {
+		select {
+		case <-o.stopping:
+			// Close the channel used to read results so that any callers waiting on
+			// results (which can happen if they submitted something before this method
+			// has a chance to execute it but after this thread pool has been stopped)
+			// so that they immediately get a zero value which they can check for.
+			close(o.res)
```
Comment (colega) [inline pkg/storegateway/indexheader/thread.go:59]: Why can't we just remove the buffer from `o.call`?

File: pkg/storegateway/indexheader/thread.go [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+		stopping: make(chan struct{}),
+		stopped:  make(chan struct{}),
+	}
+}
+
+func (o *OSThread) start() {
+	runtime.LockOSThread()
+
+	defer func() {
+		runtime.UnlockOSThread()
```
Comment (colega) [inline pkg/storegateway/indexheader/thread.go:48]: I think it's safer to remove this.

Note that:
```go
// If the calling goroutine exits without unlocking the thread,
// the thread will be terminated.
```

And also:

```go
// Before calling UnlockOSThread, the caller must ensure that the OS
// thread is suitable for running other goroutines. If the caller made
// any permanent changes to the state of the thread that would affect
// other goroutines, it should not call this function and thus leave
// the goroutine locked to the OS thread until the goroutine (and
// hence the thread) exits.
```

File: pkg/storegateway/indexheader/threadpool.go [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+	// The t.stopping channel is never written so this blocks until the channel is
+	// closed at which point the threadpool is shutting down, so we want to stop
+	// each of the expected threads in it.
+	<-t.stopping
+
+	// Stop and wait for all threads, regardless if they are "in" the pool or being
+	// used to run caller code. The avoids race conditions where threads are removed
+	// and added back to the pool while we are trying to stop all of them.
+	for _, thread := range t.threads {
+		thread.Stop()
```
Comment (colega) [inline pkg/storegateway/indexheader/threadpool.go:87]: Instead of iterating all the thread, why don't we just pass _our_ (pool's) `t.stopping` to each thread and use it as `stopping` there? That would stop all threads at the same time.

File: pkg/storegateway/indexheader/threadpool.go [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+		return nil, nil
+	}
+
+	tp := &Threadpool{
+		pool:       make(chan *OSThread, num),
+		stopping:   make(chan struct{}),
+		stopped:    make(chan struct{}),
+		numThreads: num,
+
+		timing: promauto.With(reg).NewHistogramVec(prometheus.HistogramOpts{
```
[the remark this answers is not in the document]
Comment (colega) [reply pkg/storegateway/indexheader/threadpool.go:48]: I miss the buckets here, and I think that no buckets will work well here: "waiting stage" will be something moving in milliseconds, while "complete" is something that depends a lot on a task and can be even seconds. So this will end up with a huge cardinality on the buckets.

I'd consider using a summary for the waiting time (exporting the p99, then we can graph the max of p99), and maybe (not sure about this) even dropping the "complete" stage at all since it's something out of this pool's control

File: pkg/storegateway/indexheader/threadpool.go (hunk 19 above)
Comment (56quarters) [reply pkg/storegateway/indexheader/threadpool.go:87]: Good idea, I'll give it a try!

File: pkg/storegateway/indexheader/thread.go (hunk 18 above)
Comment (56quarters) [reply pkg/storegateway/indexheader/thread.go:48]: We don't do anything in the thread that would make it unsafe to be reused by other goroutines (the only code being run in them is code that was previously being run in plain goroutines) and so I'd rather not omit a cleanup method. However, I can do this just in case if people feel strongly.

File: pkg/storegateway/indexheader/thread.go [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+	case <-o.stopping:
+		// Make sure the pool (and hence this thread) hasn't been stopped before
+		// submitting our function to run. It's still possible for the thread pool
+		// to be stopped between this check and our function actually being run by
+		// the .start() method. We handle this by closing the o.res channel as part
+		// of stopping this thread and testing for a zero value being returned
+		// (which is what reading from a closed channel does).
+		return nil, ErrPoolStopped
+	case o.call <- fn:
+		res := <-o.res
```
[the remark this answers is not in the document]
Comment (56quarters) [reply pkg/storegateway/indexheader/thread.go:93]: We can't actually use unbuffered channels:
> Actually, I don't think we can remove the buffer form o.call. If we do, the write to it in Call() will block until it's read which may not happen if the thread is in the process of stopping (i.e. .start() is no longer reading from it because o.stopping has been closed).

File: pkg/storegateway/indexheader/thread.go (hunk 17 above)
Comment (colega) [reply pkg/storegateway/indexheader/thread.go:60]: > If we do, the write to it in Call() will block until it's read which may not happen if the thread is in the process of stopping.

If the process is stopping then `Call()` will go through the `case <-o.stopping` case.

If the `Call()` function has managed to push a function through `o.call` then this goroutine will always push a result back through `o.res`.

File: pkg/storegateway/indexheader/thread.go [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+			// Close the channel used to read results so that any callers waiting on
+			// results (which can happen if they submitted something before this method
+			// has a chance to execute it but after this thread pool has been stopped)
+			// so that they immediately get a zero value which they can check for.
+			close(o.res)
+			return
+		case fn := <-o.call:
+			val, err := fn()
+			o.res <- execResult{value: val, err: err, populated: true}
+			break
```
Comment (colega) [inline pkg/storegateway/indexheader/thread.go:65]: This `break` does nothing and IMO should be removed.

File: pkg/storegateway/indexheader/thread.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (colega) [reply pkg/storegateway/indexheader/thread.go:93]: > A thread can only have tasks submitted to it from a single caller at a time, after it's been acquired via the .Call() method in ThreadPool.

If there are two threads racing, then one thread can push a call, second thread can push another call  (because channel is buffered, and the first one was already read by the thread), and then we're in a situations where both callers are racing to read a result from `o.res`, in which case it can be any of them.

```mermaid
sequenceDiagram
    participant caller1
    participant caller2
    participant o.call
    participant o.res

    caller1 ->> o.call: pushes call1
    o.call ->> thread: call1
    o.res -->> caller1: waiting for result
    caller2 ->> o.call: pushes call2
    o.res -->> caller2: waiting for result
    thread ->> o.res: res1
    o.res ->> caller2: reads res1
    o.call ->> thread: call2
    thread ->> o.res: res2
    o.res  ->> caller1: reads res2

```

File: pkg/storegateway/indexheader/thread.go (hunk 18 above)
Comment (colega) [reply pkg/storegateway/indexheader/thread.go:49]: This isn't a strong opinion, I just think that it doesn't make sense to keep a code that does nothing (in this case).

File: pkg/storegateway/indexheader/thread.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (56quarters) [reply pkg/storegateway/indexheader/thread.go:93]: This can't actually happen because functions are only pushed to `OSThread.Call` after the thread has been removed from the pool here: https://github.com/grafana/mimir/pull/1660/files#diff-74baca6d3e8d3598a308a4ec324a9e416a24a5b01c8bbd78dce0f10bdab3db49R109

Before anything is submitted to the `OSThread`, it's removed from `ThreadPool.pool` such that only a single caller of `ThreadPool.Call` is operating on it.

File: pkg/storegateway/indexheader/thread.go (hunk 17 above)
Comment (56quarters) [reply pkg/storegateway/indexheader/thread.go:60]: > If the Call() function has managed to push a function through o.call then this goroutine will always push a result back through o.res.

Yes, but the write to `o.call` will block in `OSThread.Call`.

File: pkg/storegateway/indexheader/thread.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (colega) [reply pkg/storegateway/indexheader/thread.go:93]: Can we then unexport `OSThread` to be `osThread` since it requires `ThreadPool` logic to work safely?

[the remark this answers is not in the document]
Comment (56quarters) [reply pkg/storegateway/indexheader/thread.go:93]: Yes, good idea

File: pkg/storegateway/indexheader/thread.go (hunk 17 above)
Comment (56quarters) [reply pkg/storegateway/indexheader/thread.go:60]: Turns out that the test that was hanging (because the write to `o.call` was blocking was invalid. I will update the test to match how `OSThread` is supposed to be used.

File: pkg/storegateway/indexheader/threadpool.go [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+		tp.pool <- t
+	}
+
+	return tp
+}
+
+func (t *Threadpool) run() {
+	defer func() {
+		close(t.stopped)
+	}()
```
Comment (colega) [inline pkg/storegateway/indexheader/threadpool.go:77]: Very minor nit:
```suggestion
	defer close(t.stopped)
```

File: pkg/storegateway/indexheader/thread.go [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+func (o *osThread) execute(fn func() (interface{}, error)) (interface{}, error) {
+	select {
+	case <-o.stopping:
+		// Make sure the pool (and hence this thread) hasn't been stopped before
+		// submitting our function to run.
+		return nil, ErrPoolStopped
+	case o.call <- fn:
+		res := <-o.res
+		return res.value, res.err
+	}
```
Comment (colega) [inline pkg/storegateway/indexheader/thread.go:81]: Just as a safety measure, I'd consider adding a default clause here:
```go
default:
        panic("either called on a non-running thread or used same thread for two concurrent calls")
```

This should never happen, but if we end up in this situation, it's better to panic and restart than to have an unresponsive application saying nothing.

**Review by colega — APPROVED** [review]
LGTM again, please consider the nitpicks and don't forget to update the CHANGELOG.md

File: pkg/storegateway/indexheader/thread.go (hunk 24 above)
Comment (56quarters) [reply pkg/storegateway/indexheader/thread.go:81]: OK, will do.

Comment (56quarters) [reply pkg/storegateway/indexheader/thread.go:81]: Actually, it doesn't seem like we can do this. I suspect this is because the write to `o.call` would block because the `o.start` method has not yet started consuming entries from `o.call` when run in this test and thus the `default` case gets executed. Without the `default` case, the  write to `o.call` blocks until the thread reads the value in `o.start`.

```
--- FAIL: TestOSThread_Call (0.01s)
    --- FAIL: TestOSThread_Call/run_by_thread (0.01s)
panic: either called on a non-running thread or used same thread for two concurrent calls [recovered]
        panic: either called on a non-running thread or used same thread for two concurrent calls

goroutine 8 [running]:
testing.tRunner.func1.2({0x982820, 0xc24f38})
        /usr/local/go/src/testing/testing.go:1209 +0x24e
testing.tRunner.func1()
        /usr/local/go/src/testing/testing.go:1212 +0x218
panic({0x982820, 0xc24f38})
        /usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/grafana/mimir/pkg/storegateway/indexheader.(*osThread).execute(0xc00007d400, 0xc00000ec60)
        /go/src/github.com/grafana/mimir/pkg/storegateway/indexheader/thread.go:87 +0x108
github.com/grafana/mimir/pkg/storegateway/indexheader.TestOSThread_Call.func1(0xc0000b9040)
        /go/src/github.com/grafana/mimir/pkg/storegateway/indexheader/thread_test.go:25 +0x185
testing.tRunner(0xc0000b9040, 0xa77c48)
        /usr/local/go/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:1306 +0x35a
```

**Comment by 56quarters** [comment]
Rebased and implemented the last round of review feedback.

@pracucci I did not end up implementing your suggestion to adjust `GOMAXPROCS` based on the size of the `ThreadPool` created because:
* This is an experimental feature so I think a little manual adjusting is fine for now while we're figuring out if this solves the store-gateway stability issues.
* `GOMAXPROCS` is global state and so it doesn't feel right that creating a new struct would modify it. For example, in unit tests `GOMAXPROCS` would be increased for each `ThreadPool` created.

Testing in a dev environment has shown that this change increases p99 latency for the store-gateway. We'll need to test in a higher traffic environment to determine if it prevents the issues we've been seeing with store-gateways becoming unresponsive. I think the increase in latency is OK for now since this feature is off by default and can be completely disabled. If it helps with stability, we can take another look at the added latency.

File: pkg/storegateway/indexheader/thread.go (hunk 24 above)
Comment (colega) [reply pkg/storegateway/indexheader/thread.go:83]: Oh, I see. Certainly it's hard-to-impossible to fix this (even if that goroutine is started, it may not be listening on the for-loop yet). We could set a timer here, but that would be too much hassle for something that shouldn't happen.

**Comment by pracucci** [comment]
> @pracucci I did not end up implementing your suggestion to adjust `GOMAXPROCS` based on the size of the `ThreadPool` created because:

I'm fine to postpone the discussion. I think the configuration we're adding is hard to understand for a final user (my suggestion was to attempt to simplify its usage), but I agree that given it's experimental we can postpone any discussion on how to make it easier for the final user.

File: CHANGELOG.md [hunk 25]
```diff
@@ -36,6 +36,7 @@
 * [ENHANCEMENT] Ruler: Add more detailed query information to ruler query stats logging. #1411
 * [ENHANCEMENT] Admin: Admin API now has some styling. #1482 #1549
 * [ENHANCEMENT] Alertmanager: added `insight=true` field to alertmanager dispatch logs. #1379
+* [ENHANCEMENT] Store-gateway: Add the experimental ability to run index header operations in a dedicated thread pool. #1660
```
Comment (pracucci) [inline CHANGELOG.md:39]: Please mention this feature in the list of experimental ones at `docs/sources/operators-guide/configuring/about-versioning.md`.

Comment (pracucci) [reply CHANGELOG.md:39]: In this CHANGELOG entry, please also mention the new config option introduced.

File: pkg/storegateway/indexheader/threadpool.go [hunk 26]
```diff
@@ -0,0 +1,126 @@
...
+	case thread := <-t.pool:
+		acquired := time.Now()
+
+		defer func() {
+			complete := time.Now()
+
+			t.pool <- thread
+			t.tasks.Dec()
+			t.timing.WithLabelValues(labelWaiting).Observe(start.Sub(acquired).Seconds())
+			t.timing.WithLabelValues(labelRunning).Observe(acquired.Sub(complete).Seconds())
```
Comment (pracucci) [inline pkg/storegateway/indexheader/threadpool.go:120]: These need to be inverted:

```suggestion
			t.timing.WithLabelValues(labelWaiting).Observe(acquired.Sub(start).Seconds())
			t.timing.WithLabelValues(labelRunning).Observe(complete.Sub(acquired).Seconds())
```

**Review by pracucci — APPROVED** [review]
LGTM (modulo a fix in the stats tracking)

File: pkg/storegateway/indexheader/threadpool.go (hunk 26 above)
Comment (56quarters) [reply pkg/storegateway/indexheader/threadpool.go:120]: :facepalm: will do

Comment (56quarters) [reply pkg/storegateway/indexheader/threadpool.go]: Fixed in 9df3e28966c5f6fd56f397af04358c616d7b522b

File: CHANGELOG.md (hunk 25 above)
Comment (56quarters) [reply CHANGELOG.md]: Fixed in 9df3e28966c5f6fd56f397af04358c616d7b522b
